### PR TITLE
feat(torch/nbeats): allow different sizes for backcast and forecast ,…

### PR DESCRIPTION
**Pull request by fantes**
… also implements minimal change in csvtstorchinputconn in order to do forecast of signals instead of label predicting

API : simply add `params.input.forecast_timesteps:10` and `params.input.backcast:40` this will split timeseries in sequences of length 50  and will use `backcast_timesteps` steps as input data and will output `forecast` steps as output. `labels` will be ignored.

previous API (ie lstm-like: same number of steps for input and output, `timesteps` `labels` predicted based on `timesteps` `inputs` still work)

### CODE REVIEW

File: src/backends/torch/torchinputconns.cc [hunk 1]
```diff
@@ -750,28 +753,146 @@ namespace dd
...
+                                 "discarding";
+                  }
+                else
+                  {
+                    if (static_cast<unsigned int>(vecindex) < _fnames.size())
+                      errmsg = "file " + _fnames[vecindex]
+                               + " does not contains enough timesteps, "
+                                 "discarding";
+                  }
+                _tilogger->warn(errmsg);
```
Comment (BynaryCobweb) [inline src/backends/torch/torchinputconns.cc:883]: `discard_warn` ?

File: src/backends/torch/torchinputconns.cc [hunk 2]
```diff
@@ -750,28 +753,146 @@ namespace dd
...
+        if (static_cast<unsigned int>(vecindex) < _fnames.size())
+          errmsg = "file " + _fnames[vecindex]
+                   + " does not contains enough timesteps, "
+                     "discarding";
+      }
+    _tilogger->warn(errmsg);
+  }
+
+  void CSVTSTorchInputFileConn::fill_dataset_forecast(
+      TorchDataset &dataset, const std::vector<std::vector<CSVline>> *data)
```
Comment (BynaryCobweb) [inline src/backends/torch/torchinputconns.cc:836]: It's better to pass references as parameter

File: src/backends/torch/torchinputconns.cc [hunk 3]
```diff
@@ -750,28 +753,146 @@ namespace dd
...
+        at::Tensor pst = torch::stack(pred_sequence);
+        dataset.add_batch({ dst }, { pst });
+      }
+    else // we are in inference mode, not forecast available
+      dataset.add_batch({ dst }, {});
+  }
+
+  void CSVTSTorchInputFileConn::discard_warn(
+      int vecindex, unsigned int seq_size,
+      const std::vector<std::vector<CSVline>> *data)
```
Comment (BynaryCobweb) [inline src/backends/torch/torchinputconns.cc:812]: const reference

File: src/backends/torch/torchinputconns.cc [hunk 4]
```diff
@@ -783,15 +904,59 @@ namespace dd
...
+        data_sequence.push_back(data);
+        label_sequence.push_back(label);
+      }
+    at::Tensor dst = torch::stack(data_sequence);
+    at::Tensor lst = torch::stack(label_sequence);
+    dataset.add_batch({ dst }, { lst });
+  }
+
+  void CSVTSTorchInputFileConn::fill_dataset_labels(
+      TorchDataset &dataset, const std::vector<std::vector<CSVline>> *data)
```
Comment (BynaryCobweb) [inline src/backends/torch/torchinputconns.cc:949]: const reference

File: src/backends/torch/torchinputconns.h [hunk 5]
```diff
@@ -524,10 +535,27 @@ namespace dd
...
+
+    void discard_warn(int vecindex, unsigned int seq_size,
+                      const std::vector<std::vector<CSVline>> *data);
+
+  public:
     int _offset = -1;    /**< default offset for building sequences: start of
                             sequences is at 0, offset, 2xoffset ... */
     int _timesteps = -1; /**< default empty value for timesteps */
     int _datadim = -1;   /**< default empty value for datapoints */
+    int _forecast = -1;  /**< length of forecast : will not honor labels*/
```
Comment (BynaryCobweb) [inline src/backends/torch/torchinputconns.h:558]: "labels are ignored" ?

File: tests/ut-graph.cc [hunk 6]
```diff
@@ -773,14 +773,14 @@ TEST(graphapi, complete_extract_layer_gpu)
         + str_min_vals + ",\"max_vals\":" + str_max_vals
         + "},\"output\":{}},\"data\":[\"" + csvts_predict + "\"]}";
   joutstr = japi.jrender(japi.service_predict(jpredictstr));
-  std::cout << "joutstr=" << joutstr << std::endl;
+  // std::cout << "joutstr=" << joutstr << std::endl;
```
Comment (BynaryCobweb) [inline tests/ut-graph.cc:776]: This may be useful to debug the test if it fails during CI

File: src/backends/torch/torchinputconns.cc (hunk 1 above)
Comment (fantes) [reply src/backends/torch/torchinputconns.cc:883]: cleanup up, thx

File: src/backends/torch/torchinputconns.cc (hunk 2 above)
Comment (fantes) [reply src/backends/torch/torchinputconns.cc:836]: ok, done

File: src/backends/torch/torchinputconns.cc (hunk 3 above)
Comment (fantes) [reply src/backends/torch/torchinputconns.cc:812]: ok, done

File: src/backends/torch/torchinputconns.h (hunk 5 above)
Comment (fantes) [reply src/backends/torch/torchinputconns.h:558]: ok

File: tests/ut-graph.cc (hunk 6 above)
Comment (fantes) [reply tests/ut-graph.cc:776]: ok

File: src/backends/torch/torchinputconns.cc (hunk 3 above)
Comment (fantes) [reply src/backends/torch/torchinputconns.cc:812]: cleanup up a lot

File: src/backends/torch/torchinputconns.cc (hunk 4 above)
Comment (fantes) [reply src/backends/torch/torchinputconns.cc:949]: again, cleaned up -> removed

**Review by BynaryCobweb — APPROVED** [review]

**Comment by fantes** [comment]
api changed from `timesteps` + `forecast` to `backcast_timesteps` + `forecast_timesteps`

**Review by beniz — APPROVED** [review]

**Pull request merged**